torch/csrc/jit/fusion_compiler.cpp

```diff
@@ -1047,67 +1140,409 @@ struct CPUFusionFunction : public CompiledFusionFunction {
+//   - a.s when a is a tensor is a shorthand for a.shape.
+//   - B is a shorthand for the broadcasting/expanding function. It is used as a
+//     vararg function.
+//   - E is a shorthand for expand function.
+//   - Every pointwise operation can be equivalently rewritten as
+//     f(a, b) = f^(E(a, B(a.s, b.s)), E(b, B(a.s, b.s))),
+//     where f^ is a non-broadcasting verison of f.
+//   - A set of inputs that are used to produce a certain graph output is referred to
+//     as the output's broadcasting group (see Lemma 2. for explanation why).
+//
+// Lemma 1. Set of lists of integers (shapes) + { _|_ (bottom/error marker) }, with the
+//          operation of broadcasting (returning bottom upon shape mismatch) forms a monoid.
+//          In simpler terms: broadcasting is associative, i.e. B(a, B(b, c)) == B(B(a, b), c).
+//
+// Proof.   Satisfies all monoid laws:
+//            - Closed under broadcasting (trivial)
+//            - Empty shape is the identity element: B(a, []) == B([], a) == a
+//            - Associativity: A simple visual proof is that you can expand 3 tensors
+//                at the same time by stacking their sizes (with alignment to the right),
+//                just as you'd do in the case of 2 tensors, but with an intermediate
+//                (the algorithm ends up being pretty much the same).
+//
+// Lemma 2. Shape of an output of an arbitrary DAG of pointwise ops depends only on the set
+//          of inputs used in this DAG and is equal to B([i.shape for i in used_inputs]).
+//
+// Proof.   Let G be any DAG of pointwise ops and < be any valid topological
+//          ordering on nodes of G. Proof by induction over <.
+//          Base case (graph input):
+//            Trivial (input is also an output).
+//          Step (n = f(q, r)):
+//            Let QS (RS) be the set of shapes of inputs that q (r) depends on.
+//            Note that the set of inputs that n depends on is exactly QS + RS.
+//            shape(n) == shape(f(q, r))
+//                          (def of f)
+//                     == shape(f^(E(q, B(q.s, r.s)), E(r, B(q.s, r.s))))
+//                          (output shape of f^ is equal to either of argument shapes)
+//                     == shape(E(q, B(q.s, r.s)))
+//                          (property of expand)
+//                     == B(q.s, r.s)
+//                          (induction assumption)
+//                     == B(B(QS...), B(RS...))
+//                          (Lemma 1.)
+//                     == B(QS..., RS...)
+//                          (repeated shapes don't matter for broadcasting)
+//                     == B((QS + RS)...)
+//
+// Lemma 3. Expands are distributive over pointwise ops, i.e. E(f(a, b), s) = f(E(a, s), E(b, s))
+// Lemma 4. Expands can be collapsed, i.e. E(E(x, s1), s2) = E(x, B(s1, s2)).
+// Proof.   A simple exercise for the reader :)
+//
+// Theorem. If all (pre-concat-)outputs have equal shapes, then we can push the expands to
```

Comment:
In general you can, but you can't if you want to use the fused kernel. In this case, it would fail the check and use the fallback, which can operate on arbitrary shapes.

Reply:
We had an offline conversation and I was reminded why we have this invariant and how concat works (modeled as "multiple outputs"). Would be nice if there was a comment about it somewhere...